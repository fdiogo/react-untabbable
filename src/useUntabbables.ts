import { Options } from './untabbable';
import useUntabbable from './useUntabbable';

function useUntabbables(refs: [React.RefObject<HTMLElement>], options: Options = {}) {
    refs.forEach(ref => { useUntabbable(ref, options) });
}

export default useUntabbables;
