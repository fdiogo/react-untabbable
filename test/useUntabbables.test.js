import React from 'react';
import { render, cleanup } from 'react-testing-library';

import useUntabbables from '../src/useUntabbables';

afterEach(cleanup);

// We mock tabbable here since what we are really interested on
// testing is the toggling of the tabindex.
jest.mock('tabbable', () => (element, options) => {
    const tabbableElements = [];

    if (options.includeContainer) {
        tabbableElements.push(element);
    }

    tabbableElements.push(...element.children);

    return tabbableElements;
});

const HookWrapper = (props) => {
    const { children, targetRefs, options } = props;

    useUntabbables(targetRefs, options)

    return children;
};

const n = 5;

describe('useUntabbable', () => {
    it('should set the focusable children with tabindex=-1', () => {
        const refs = Array.from({ length: n }, () => React.createRef());

        const { container, getAllByText } = render(
            <HookWrapper targetRefs={refs}>
                {
                    refs.map((ref, i) => (
                        <div key={i} ref={ref}>
                            <button>First</button>
                            <button>Second</button>
                        </div>
                    ))
                }
            </HookWrapper>
        );

        const firstButtons = getAllByText('First');
        const secondButtons = getAllByText('Second');

        firstButtons.forEach((firstButton) => {
            expect(firstButton.getAttribute("tabindex")).toBe("-1");
        });

        secondButtons.forEach((secondButton) => {
            expect(secondButton.getAttribute("tabindex")).toBe("-1");
        });

        [...container.children].forEach(child => {
            expect(child.getAttribute("tabindex")).not.toBe("-1");
        });

        expect(container).toMatchSnapshot();
    });

    it('should restore the previous tabindex when disabled', () => {
        const refs = Array.from({ length: n }, () => React.createRef());

        const { container, getAllByText } = render(
            <HookWrapper targetRefs={refs}>
                {
                    refs.map((ref, i) => (
                        <div key={i} ref={ref}>
                            <button tabIndex="2">First</button>
                            <button>Second</button>
                        </div>
                    ))
                }
            </HookWrapper>
        );

        const firstButtons = getAllByText('First');
        const secondButtons = getAllByText('Second');

        firstButtons.forEach((firstButton) => {
            expect(firstButton.getAttribute("tabindex")).toBe("-1");
        });

        secondButtons.forEach((secondButton) => {
            expect(secondButton.getAttribute("tabindex")).toBe("-1");
        });

        render(
            <HookWrapper targetRefs={refs} options={{ disabled: true }}>
                {
                    refs.map((ref, i) => (
                        <div key={i} ref={ref}>
                            <button tabIndex="2">First</button>
                            <button>Second</button>
                        </div>
                    ))
                }
            </HookWrapper>,
            { container });

        firstButtons.forEach((firstButton) => {
            expect(firstButton.getAttribute("tabindex")).toBe("2");
        });

        secondButtons.forEach((secondButton) => {
            expect(secondButton.getAttribute("tabindex")).toBe(null);
        });

        expect(container).toMatchSnapshot();
    });

    it('should also disable the container when `includeContainer` is true', () => {
        const refs = Array.from({ length: n }, () => React.createRef());

        const { container, getAllByText } = render(
            <HookWrapper targetRefs={refs} options={{ includeContainer: true }}>
                {
                    refs.map((ref, i) => (
                        <div key={i} ref={ref}>
                            <button>First</button>
                            <button>Second</button>
                        </div>
                    ))
                }
            </HookWrapper>
        );

        const firstButtons = getAllByText('First');
        const secondButtons = getAllByText('Second');

        firstButtons.forEach((firstButton) => {
            expect(firstButton.getAttribute("tabindex")).toBe("-1");
        });

        secondButtons.forEach((secondButton) => {
            expect(secondButton.getAttribute("tabindex")).toBe("-1");
        });

        [...container.children].forEach(child => { expect(child.getAttribute("tabindex")).toBe("-1"); })
        expect(container).toMatchSnapshot();
    });

    it('should not alter the tabindex when `disabled` is true', () => {
        const refs = Array.from({ length: n }, () => React.createRef());

        const { container, getAllByText } = render(
            <HookWrapper targetRefs={refs} options={{ disabled: true }}>
                {
                    refs.map((ref, i) => (
                        <div key={i} ref={ref}>
                            <button>First</button>
                            <button>Second</button>
                        </div>
                    ))
                }
            </HookWrapper>
        );

        const firstButtons = getAllByText('First');
        const secondButtons = getAllByText('Second');


        firstButtons.forEach((firstButton) => {
            expect(firstButton.getAttribute("tabindex")).not.toBe("-1");
        });

        secondButtons.forEach((secondButton) => {
            expect(secondButton.getAttribute("tabindex")).not.toBe("-1");
        });

        [...container.children].forEach(child => { expect(child.getAttribute("tabindex")).not.toBe("-1"); })
        expect(container).toMatchSnapshot();
    });
});