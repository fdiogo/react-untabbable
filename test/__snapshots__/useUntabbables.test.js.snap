// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUntabbable should also disable the container when \`includeContainer\` is true 1`] = `
<div>
  <div
    tabindex="-1"
  >
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div
    tabindex="-1"
  >
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div
    tabindex="-1"
  >
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div
    tabindex="-1"
  >
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div
    tabindex="-1"
  >
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
</div>
`;

exports[`useUntabbable should not alter the tabindex when \`disabled\` is true 1`] = `
<div>
  <div>
    <button>
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button>
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button>
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button>
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button>
      First
    </button>
    <button>
      Second
    </button>
  </div>
</div>
`;

exports[`useUntabbable should restore the previous tabindex when disabled 1`] = `
<div>
  <div>
    <button
      tabindex="2"
    >
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="2"
    >
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="2"
    >
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="2"
    >
      First
    </button>
    <button>
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="2"
    >
      First
    </button>
    <button>
      Second
    </button>
  </div>
</div>
`;

exports[`useUntabbable should set the focusable children with tabindex=-1 1`] = `
<div>
  <div>
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
  <div>
    <button
      tabindex="-1"
    >
      First
    </button>
    <button
      tabindex="-1"
    >
      Second
    </button>
  </div>
</div>
`;
